### PR TITLE
Replace asynctask with coroutines

### PR DESCRIPTION
**Pull request by ShivamPokhriyal**
## Summary

This PR introduced a `CoroutinesAsyncTask` class which is a replacement for the asynctask. It uses the same `threadpoolexecutor` from asynctask as a coroutine dispatcher. 

## Safety Assurance

- [x] If the PR is high risk, "High Risk" label is set
- [x] I have confidence that this PR will not introduce a regression for the reasons below
- [x] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly

### CODE REVIEW

**Comment by dimagimon** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR does not contain any JIRA issue keys in the PR title or commit messages (e.g. KEY-123)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
@shubham1g5 bumping this for review.

**Review by shubham1g5 — COMMENTED** [review]
@ShivamPokhriyal Looks quite solid to me. Thanks for taking up this work. I am curious having done this work if you see any red flags with this migration and if you have thoughts on what should we test for in QA to ensure that we don't intorduce major issues.

File: app/src/org/commcare/tasks/templates/CoroutineAsyncTask.kt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import androidx.annotation.WorkerThread
+import kotlinx.coroutines.*
+import kotlinx.coroutines.CancellationException
+import org.commcare.CommCareApplication
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * @author $|-|!˅@M
+ */
+abstract class CoroutinesAsyncTask<Params, Progress, Result> {
```
[the remark this answers is not in the document]
Comment (ShivamPokhriyal) [reply app/src/org/commcare/tasks/templates/CoroutineAsyncTask.kt:13]: Mostly this https://medium.com/@rahulraks25/migrating-asynctask-to-kotlin-coroutines-e93e9f45d995, but I kinda made some modifications to allows both serial and parallel executions. 
Will add a comment.

**Comment by ShivamPokhriyal** [comment]
> if you see any red flags with this migration and if you have thoughts on what should we test for in QA to ensure that we don't introduce any major issues.

This is a really great question. I think our instrumentation tests provide a pretty good coverage for background tasks executions which is kinda why I used an almost similar executor configuration in test https://github.com/dimagi/commcare-android/pull/2524/files#diff-17dd0e65d12c5c9c80fea35e8f44776da3dc86037aec606dcf1457c8daa975fdR28. 

But espresso makes sure to wait until all background execution is completed(which isn't an ideal real world scenario) and robolectric requires tests to run on a single thread(which is why I simply replaced it with main dispatcher), so there's still a possibility that this change can cause a bug which would become hard to track later. 

So it definitely does require manual QA. I think I can talk to QA to do some tests around this PR(just wanted to get all the reviews done before doing QA so that QA gets the up to date apk) and one of the things we can explore in tests is around multiple quick form submissions while toggling the internet. I will probably more details on the ticket and will ping you on it.

**Comment by shubham1g5** [comment]
Thanks @ShivamPokhriyal . That makes sense to me. I am also worried about issues related to memory with this change. Would it be possible for you to do some memory profiling for this change as well and see if there is a significant memory surge in either direction when comparing to current Async task code. 

One other thing, You mentioned that AS doesn't have good support for debugging coroutines. Does this migration make it harder for us to maintain and debug code long term ? I think it should be fine though just want to get your perspective.

**Comment by ShivamPokhriyal** [comment]
> Would it be possible for you to do some memory profiling for this change as well and see if there is a significant memory surge in either direction when comparing to current Async task code.

Sure will do that. 


> One other thing, You mentioned that AS doesn't have good support for debugging coroutines. Does this migration make it harder for us to maintain and debug code long term ? I think it should be fine though just want to get your perspective.

What I meant earlier was that espresso doesn't support coroutines the same way it supports async tasks. So I had to mock the dispatcher with a test one(although I tried to provide the similar configuration). Right now the change works and the tests are passing, but that api I used was still in beta and may change in future. 
Apart from this, I don't see any other concern with this migration.

**Comment by ShivamPokhriyal** [comment]
Did some profiling, memory consumption before and after the removal of asynctask looks almost similar.

**Comment by ShivamPokhriyal** [comment]
Weird, although `ci skip` marks the build as `NOT_BUILT`, but still the check is marked as failure.

**Comment by ShivamPokhriyal** [comment]
Removed the ci skip plugin. 
@damagatchi retest this please.

**Comment by ShivamPokhriyal** [comment]
Fixed the build error https://github.com/dimagi/commcare-core/pull/1041
@damagatchi retest this please.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

**Comment by ShivamPokhriyal** [comment]
Closing this PR since I'm not able to push changes directly. Will create a new PR from my fork.

**Pull request closed without merging**